### ISSUE TITLE
Could you please share your testing code.

### ISSUE THREAD

**Issue by zoumaxingjiulic** [neel-dey/Atlas-GAN#5]
Hi @neel-dey ,

First of all, thanks for making the code available.Could you please share your testing code with me, or alternatively, could you send it to my email? My email address is wcc15966320402@gmail.com. Your assistance is greatly appreciated!

Thank you sincerely.

**Comment by neel-dey** [comment]
Hi @zoumaxingjiulic ,

Sorry, the testing code for this paper was just several dozen one-off scripts and wasn't systematized to be in a shareable format (due to the last minute rush for the paper deadline). I'd be happy to help you recreate it though.

If you're interested in Table 1, this is how we did it:
- **Dice**: We had segmentations for all the input images and we segmented the atlases, see Appendices B.1.1 and B.1.1.2 for how we did it (you can just use [SynthSeg](https://surfer.nmr.mgh.harvard.edu/fswiki/SynthSeg) to segment the atlases as well now, if you're using adult neuroimages). For testing subjects, we do a forward pass through the registration network with the atlas and the subject, and we take the displacement field from [this line](https://github.com/neel-dey/Atlas-GAN/blob/main/src/networks.py#L438) of the network. You then apply the field to the segmentations as  `atlas_to_subject_seg = SpatialTransformer('nearest')([atlas_seg, diff_field])` and then just compute Dice between `atlas_to_subject_seg` and `subject_seg`.
- **Mov. Def**: After training, we just take the norm of the moving average, stored as weights of the model in [this layer](https://github.com/neel-dey/Atlas-GAN/blob/main/src/networks.py#L433C23-L433C23).
- **EFC**: We used [this implementation](https://mriqc.readthedocs.io/en/latest/_modules/mriqc/qc/anatomical.html#efc). Make sure to use a foreground mask and use the same mask for all the conditional templates for a method, as EFC is quite sensitive to these modeling decisions.
- **Avg. JacDet**: We computed the warps between all the testing images and their corresponding templates and then computed the determinant of Jacobian with [this function](https://github.com/voxelmorph/voxelmorph/blob/ca3d47a2c254aae9a0c0e1b30c24c324c211ebc8/voxelmorph/py/utils.py#L473).
- **Avg. Def**: We computed the warps between all the testing images and their corresponding templates and then computed the average L2 norm of the per-subject displacement fields.

Do you have any other questions?

Best,
Neel

**Comment by zoumaxingjiulic** [comment]
Hi @neel-dey  ,

Thank you very much for your reply! Your response was incredibly detailed, and I believe I have a general understanding of what you explained. Allow me to ask a few more questions:

1.When debugging the code (I would like to examine certain details in the "film" function), I am unable to debug into the internals of the function. This may be related to the code structure of TensorFlow's computation graph. Any insights on how to address this would be greatly appreciated.

2.I noticed the mention of "losswt_tv" in your code, although it wasn't used in the article. Would incorporating "losswt_tv" contribute to making the constructed model more perfect?

3.Could you provide me with some of the test scripts? I am interested in understanding how the model is loaded, especially since I have been using PyTorch.

Once again, I appreciate your thoughtful responses, and I am looking forward to hearing from you!

**Comment by neel-dey** [comment]
1. Assuming that you want to inspect the intermediate outputs interactively during the training loop: Yeah, to optimally use the GPU during training, we use graph mode and that's not interactively debuggable. To enable eager execution and debug, search in the codebase for the `@tf.function` decorator above functions and comment out those lines. [Here](https://github.com/search?q=repo%3Aneel-dey%2FAtlas-GAN%20tf.function&type=code) they are. FYI, pytorch now has this eager/compiled distinction with `torch.compile` too :)

2. That's from an old version of the framework when the GAN loss was being unstable. After figuring out all the stabilization, we didn't find it to be necessary and decided to leave it out for training speed. There are follow-up papers that use it though, so you could try it. We found a version of it beneficial in some more recent [preliminary atlas work](https://arxiv.org/pdf/2311.02874.pdf).

3. Model loading is done exactly as in [these lines](https://github.com/neel-dey/Atlas-GAN/blob/main/train_script.py#L280-L293), i.e. initialize the optimizers and models, create the checkpoint, and then use `checkpoint.restore(...).assert_consumed()`

FYI, in [this short paper](https://arxiv.org/pdf/2311.02874.pdf) (mentioned above), we make some of its testing code available in pytorch. Most of the evaluation strategy is similar, but we drop the requirement to segment the atlas for Dice evaluation and instead use the ["atlas as bridge"](https://arxiv.org/abs/2202.03563) framework.

**Comment by zoumaxingjiulic** [comment]
Thank you for your response. Your answer has resolved my issue. Once again, I appreciate your assistance.

**Comment by neel-dey** [comment]
Happy to help! Feel free to open a new issue or email if you have any other questions.
